### ISSUE TITLE
maybe add aitchison distance

### ISSUE THREAD

**Issue by argriffing** [scipy/scipy#4480]
This is a distance between probability distributions, or more generally between points in the interior of a simplex.
https://mail.python.org/pipermail/python-list/2006-April/423384.html

There are so many scipy distance functions, maybe it already exists under another name?

**Comment by josef-pkt** [comment]
If you want those as transformations, then they would be similar to a multivariate extension of `expit` and similar, and could go next to those.

**Comment by argriffing** [comment]
I just tried to add this as a distance, but I see that it's not so easy.  There are a bunch of places in the code that would have to be changed, and the distances framework seems to expect some C code to be written when a new distance function is added.

If I remember correctly, the distances framework also adds a function that computes all pairwise distances between points.  But for the Aitchison distance this could be done more efficiently by transforming the points once (using the ilr transform) and then just computing the pairwise Euclidean distances.  On the other hand I've never used that particular feature of scipy.spatial.distance functions.

**Comment by essicolo** [comment]
It could just be added to `scipy.spatial.distance` as:

```
def aitchison(u, v):
    """
    Computes the Aitchison distance between two 1-D arrays.
    The Aitchison distance between 1-D arrays `u` and `v`, is defined as
    .. math::
       \sqrt{\frac{1}{2D}\Sigma_{i=1}^D\Sigma_{j=1}^D\left(\log\left(\frac{u_i}{u_j}\right)-\log\left(\frac{v_i}{v_j}\right)\right)}
    Parameters
    ----------
    u : (N,) array_like
            Input array.
    v : (N,) array_like
        Input array.
    Returns
    -------
    aitchison : double
        The Aitchison distance between vectors `u` and `v`.
    """
    
    u = _validate_vector(u)
    v = _validate_vector(v)
    
    D = u.shape[0]
    sum_i = 0
    for i in range(0, D):
        sum_j = 0
        for j in range(0, D):
            sum_j = sum_j + (np.log(u[i]/u[j]) - np.log(v[i]/v[j]))**2
        sum_i = sum_i + sum_j
    
    return(np.sqrt(1/(2*D) * sum_i))
```

What do you think?

**Comment by jaimefrio** [comment]
Do you lose numerical stability if you work out the math for `clr_u - clr_v`? If I'm doing my math correctly your calculation is the same as:

    log_u_v = np.log(u / v)
    dist = norm(log_u_v - np.mean(log_u_v)))

**Comment by essicolo** [comment]
Great catch! AFAIK your formulation is equivalent, but simpler. I would also add a validation to raise an error when values are not positive.

```
def _validate_aitchison(w, u_v):
    if np.any(w <= 0):
        raise ValueError('Values in %s should be positive' % u_v)
    return(w)

def aitchison(u, v):
    """
    Computes the Aitchison distance between two 1-D arrays.
    The Aitchison distance between 1-D arrays `u` and `v`, is defined as
    .. math::
       {||log \left( \frac{u}{v} \right) - mean \left(log \left( \frac{u}{v} \right) \right) ||}_2
       
    Parameters
    ----------
    u : (N,) array_like
        Input array.
    v : (N,) array_like
        Input array.
    Returns
    -------
    aitchison : double
        The Aitchison distance between vectors `u` and `v`.
    """
    
    u = _validate_vector(u)
    v = _validate_vector(v)
    _validate_aitchison(u, 'u')
    _validate_aitchison(v, 'v')
    log_u_v = np.log(u / v)
    dist = norm(log_u_v - np.mean(log_u_v))
    return(dist)
```

Will you pull a request?

**Comment by prubbens** [comment]
Can't seem to find this back in the scipy library. Is this implemented by now?

**Comment by rkern** [comment]
No, if you follow the linked PR, you can see that it was closed without merging because it's better for the user to transform the data and then calculate the Euclidean distance.

**Comment by mortonjt** [comment]
@rkern , @essicolo just become it is better for the user to transform the data, doesn't mean that there  shouldn't be an implementation in scipy right?  It just means that the implementation is that much easier.

It would be _huge_ benefit to the overall scientific community if this is more readily available.  I can literally point to hundreds of papers pulling out patterns with Aitchison distance that are completely missed by other metrics.

**Comment by rgommers** [comment]
@mortonjt good point. The discussion on gh-6983 wasn't really a decision that it doesn't make sense to add the Aitchison distance. @essicolo just closed it, after a suggestion that seems to have been meant to add it without having to do a C implementation.

**Comment by jolespin** [comment]
Is this available in the current stable version?

**Comment by jolespin** [comment]
@mortonjt do you know if there are plans to have this in scikitbio if it does not appear in scipy?

**Comment by mortonjt** [comment]
We don't have any plans to add this into skbio in the near future.

But this is super easy to code up -- its essentially
```python
from skbio.stats.composition import clr
from scipy.spatial.distance import euclidean
def aitchison(x, y):
    return euclidean(clr(x), clr(y))
```

That is how it has been implemented in qiime2 (see here:
https://github.com/qiime2/q2-diversity/blob/master/q2_diversity/_beta/_method.py#L116
)

Note that this is not great when there are a lot of zeros in the data - so
I would defer to robust aitchison pca to handle this
https://github.com/biocore/DEICODE
